### ISSUE TITLE
[Bug] Backspace in unsigned book crashing client

### ISSUE THREAD

**Issue by mrfloris** [Zrips/CMI#2603]
**Description of issue:**

When it does not happen:
Create a fresh new 1.13.2 spigot world, with 0 plugins. 
Make a book/quill filled with random text. Done (but dont sign it)
Upgrade with --forceUpgrade to 1.15.1 and start the server. 
Open the book and press backspace: it works, nothing happens. 

Create a fresh new 1.13.2 spigot world, with 0 plugins. 
Make a book/quill filled with random text. Done (but dont sign it)
Upgrade with --forceUpgrade to 1.15.1 and start the server (with CMI installed)
Open the book and press backspace: it works, nothing happens. 

When it does happen:
Create a fresh new 1.13.2 spigot world, with 1 plugin: CMI
Make a book/quill filled with random text. Done (but dont sign it)
Upgrade with --forceUpgrade to 1.15.1 and start the server. (with CMI installed)
Open the book and press backspace: now it does not work. It almost instantly crashes.

I've tested this 3x to try and crash without cmi plugin, trying to convert various times from 1.13.2 with and with plugin and to 1.15.1 with and without plugin. 

And when CMI is installed on 1.13.2 and 1.15.1 before and after upgrade. And you write book with cmi installed, and after upgrade open it and press backspace in it a few times: poof, client crash.

---

**ERROR (DELETE IF YOU HAVE NO ERROR):**
```
TWO DIFFERENT PLAYERS, TWO DIFFERENT SYSTEMS:

---- Minecraft Crash Report ----
// Who set us up the TNT?

Time: 30/12/19 17:30
Description: keyPressed event handler

java.lang.StringIndexOutOfBoundsException: String index out of range: 255
	at java.lang.AbstractStringBuilder.deleteCharAt(AbstractStringBuilder.java:797)
	at java.lang.StringBuilder.deleteCharAt(StringBuilder.java:253)
	at dgy.b(SourceFile:323)
	at dgy.b(SourceFile:279)
	at dgy.keyPressed(SourceFile:230)
	at dbj.a(SourceFile:336)
	at dbj$$Lambda$2567/985564948.run(Unknown Source)
	at dfz.wrapScreenError(SourceFile:447)
	at dbj.a(SourceFile:334)
	at dbj.c(SourceFile:429)
	at dbj$$Lambda$2561/275407904.run(Unknown Source)
	at aip.execute(SourceFile:94)
	at dbj.b(SourceFile:429)
	at dbj$$Lambda$1500/1401226549.invoke(Unknown Source)
	at org.lwjgl.glfw.GLFWKeyCallbackI.callback(GLFWKeyCallbackI.java:37)
	at org.lwjgl.system.JNI.invokeV(Native Method)
	at org.lwjgl.glfw.GLFW.glfwPollEvents(GLFW.java:3101)
	at com.mojang.blaze3d.systems.RenderSystem.flipFrame(SourceFile:105)
	at cxu.e(SourceFile:301)
	at dbl.d(SourceFile:970)
	at dbl.b(SourceFile:585)
	at net.minecraft.client.main.Main.main(SourceFile:202)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at java.lang.AbstractStringBuilder.deleteCharAt(AbstractStringBuilder.java:797)
	at java.lang.StringBuilder.deleteCharAt(StringBuilder.java:253)
	at dgy.b(SourceFile:323)
	at dgy.b(SourceFile:279)
	at dgy.keyPressed(SourceFile:230)
	at dbj.a(SourceFile:336)
	at dbj$$Lambda$2567/985564948.run(Unknown Source)

-- Affected screen --
Details:
	Screen name: dgy
Stacktrace:
	at dfz.wrapScreenError(SourceFile:447)
	at dbj.a(SourceFile:334)
	at dbj.c(SourceFile:429)
	at dbj$$Lambda$2561/275407904.run(Unknown Source)
	at aip.execute(SourceFile:94)
	at dbj.b(SourceFile:429)
	at dbj$$Lambda$1500/1401226549.invoke(Unknown Source)
	at org.lwjgl.glfw.GLFWKeyCallbackI.callback(GLFWKeyCallbackI.java:37)
	at org.lwjgl.system.JNI.invokeV(Native Method)
	at org.lwjgl.glfw.GLFW.glfwPollEvents(GLFW.java:3101)
	at com.mojang.blaze3d.systems.RenderSystem.flipFrame(SourceFile:105)

-- Affected level --
Details:
	All players: 1 total; [dpy['the456gamer'/31091, l='MpServer', x=19.44, y=57.00, z=57.33]]
	Chunk stats: Client Chunk Cache: 289, 169
	Level dimension: minecraft:the_end
	Level name: MpServer
	Level seed: 851454124252874499
	Level generator: ID 00 - default, ver 1. Features enabled: false
	Level generator options: {}
	Level spawn location: World: (-8215,89,6406), Chunk: (at 9,5,6 in -514,400; contains blocks -8224,0,6400 to -8209,255,6415), Region: (-17,12; contains chunks -544,384 to -513,415, blocks -8704,0,6144 to -8193,255,6655)
	Level time: 3870250 game time, 3870250 day time
	Level storage version: 0x00000 - Unknown?
	Level weather: Rain time: 0 (now: false), thunder time: 0 (now: false)
	Level game mode: Game mode: survival (ID 0). Hardcore: false. Cheats: false
	Server brand: Spigot
	Server type: Non-integrated multiplayer server
Stacktrace:
	at dnl.a(SourceFile:450)
	at dbl.c(SourceFile:1878)
	at dbl.b(SourceFile:600)
	at net.minecraft.client.main.Main.main(SourceFile:202)

-- System Details --
Details:
	Minecraft Version: 1.15.1
	Minecraft Version ID: 1.15.1
	Operating System: Windows 10 (amd64) version 10.0
	Java Version: 1.8.0_51, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
	Memory: 1515050888 bytes (1444 MB) / 1979711488 bytes (1888 MB) up to 2147483648 bytes (2048 MB)
	CPUs: 8
	JVM Flags: 9 total; -XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump -Xss1M -Xmx2G -XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32M
	Launched Version: 1.15.1
	Backend library: LWJGL version 3.2.2 build 10
	Backend API: Radeon RX 570 Series GL version 4.6.13559 Compatibility Profile Context 26.20.12028.2, ATI Technologies Inc.
	GL Caps: Using framebuffer using OpenGL 3.0
	Using VBOs: Yes
	Is Modded: Probably not. Jar signature remains and client brand is untouched.
	Type: Client (map_client.txt)
	Resource Packs: 
	Current Language: English (US)
	CPU: 8x Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz


	---- floris:


---- Minecraft Crash Report ----
// This is a token for 1 free hug. Redeem at your nearest Mojangsta: [~~HUG~~]

Time: 12/30/19 6:35 PM
Description: keyPressed event handler

java.lang.StringIndexOutOfBoundsException: String index out of range: 255
	at java.lang.AbstractStringBuilder.deleteCharAt(AbstractStringBuilder.java:797)
	at java.lang.StringBuilder.deleteCharAt(StringBuilder.java:253)
	at dgy.b(SourceFile:323)
	at dgy.b(SourceFile:279)
	at dgy.keyPressed(SourceFile:230)
	at dbj.a(SourceFile:336)
	at dfz.wrapScreenError(SourceFile:447)
	at dbj.a(SourceFile:334)
	at dbj.c(SourceFile:429)
	at aip.execute(SourceFile:94)
	at dbj.b(SourceFile:429)
	at org.lwjgl.glfw.GLFWKeyCallbackI.callback(GLFWKeyCallbackI.java:37)
	at org.lwjgl.system.JNI.invokeV(Native Method)
	at org.lwjgl.glfw.GLFW.glfwPollEvents(GLFW.java:3050)
	at com.mojang.blaze3d.systems.RenderSystem.flipFrame(SourceFile:98)
	at cxu.e(SourceFile:301)
	at dbl.d(SourceFile:970)
	at dbl.b(SourceFile:585)
	at net.minecraft.client.main.Main.main(SourceFile:202)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at java.lang.AbstractStringBuilder.deleteCharAt(AbstractStringBuilder.java:797)
	at java.lang.StringBuilder.deleteCharAt(StringBuilder.java:253)
	at dgy.b(SourceFile:323)
	at dgy.b(SourceFile:279)
	at dgy.keyPressed(SourceFile:230)
	at dbj.a(SourceFile:336)

-- Affected screen --
Details:
	Screen name: dgy
Stacktrace:
	at dfz.wrapScreenError(SourceFile:447)
	at dbj.a(SourceFile:334)
	at dbj.c(SourceFile:429)
	at aip.execute(SourceFile:94)
	at dbj.b(SourceFile:429)
	at org.lwjgl.glfw.GLFWKeyCallbackI.callback(GLFWKeyCallbackI.java:37)
	at org.lwjgl.system.JNI.invokeV(Native Method)
	at org.lwjgl.glfw.GLFW.glfwPollEvents(GLFW.java:3050)
	at com.mojang.blaze3d.systems.RenderSystem.flipFrame(SourceFile:98)

-- Affected level --
Details:
	All players: 1 total; [dpy['mrfloris'/15066, l='MpServer', x=-264.70, y=78.00, z=187.30]]
	Chunk stats: Client Chunk Cache: 289, 147
	Level dimension: minecraft:overworld
	Level name: MpServer
	Level seed: 3390704452293174725
	Level generator: ID 00 - default, ver 1. Features enabled: false
	Level generator options: {}
	Level spawn location: World: (56,85,-74), Chunk: (at 8,5,6 in 3,-5; contains blocks 48,0,-80 to 63,255,-65), Region: (0,-1; contains chunks 0,-32 to 31,-1, blocks 0,0,-512 to 511,255,-1)
	Level time: 2582174317 game time, 2530346033 day time
	Level storage version: 0x00000 - Unknown?
	Level weather: Rain time: 0 (now: true), thunder time: 0 (now: false)
	Level game mode: Game mode: creative (ID 1). Hardcore: false. Cheats: false
	Server brand: Spigot
	Server type: Non-integrated multiplayer server
Stacktrace:
	at dnl.a(SourceFile:450)
	at dbl.c(SourceFile:1878)
	at dbl.b(SourceFile:600)
	at net.minecraft.client.main.Main.main(SourceFile:202)

-- System Details --
Details:
	Minecraft Version: 1.15.1
	Minecraft Version ID: 1.15.1
	Operating System: Mac OS X (x86_64) version 10.15.2
	Java Version: 1.8.0_74, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
	Memory: 463668008 bytes (442 MB) / 1073741824 bytes (1024 MB) up to 8589934592 bytes (8192 MB)
	CPUs: 8
	JVM Flags: 8 total; -Xss1M -Xmx8G -XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32M
	Launched Version: 1.15.1
	Backend library: LWJGL version 3.2.1 build 12
	Backend API: AMD Radeon Pro 580 OpenGL Engine GL version 2.1 ATI-3.4.19, ATI Technologies Inc.
	GL Caps: Using framebuffer using ARB_framebuffer_object extension
	Using VBOs: Yes
	Is Modded: Probably not. Jar signature remains and client brand is untouched.
	Type: Client (map_client.txt)
	Resource Packs: vanilla
	Current Language: English (US)
	CPU: 8x Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
```

**CONFIG SECTION (DELETE IF NOT RELEVANT):**
```
default installation, no config files touched. 
```

---

**Cmi Version (using`/cmi version`):** 
```
>cmi version
[19:31:37] [Server thread/INFO]: --------------------------------------------------
[19:31:37] [Server thread/INFO]: CMI: 8.6.16.5
[19:31:37] [Server thread/INFO]: Server version: Spigot 1.15.1-R0.1-SNAPSHOT
[19:31:37] [Server thread/INFO]: --------------------------------------------------
>
```

**Server Type (Spigot/Paperspigot/etc):**  SPIGOT 1.15.1 (created today with buildtools on the same system)

**Server Version (using `/ver`):** 
```
>ver
[19:31:22] [Server thread/INFO]: This server is running CraftBukkit version git-Spigot-492a779-731e819 (MC: 1.15.1) (Implementing API version 1.15.1-R0.1-SNAPSHOT)
[19:31:22] [Server thread/INFO]: Checking version, please wait...
[19:31:23] [Thread-14/INFO]: You are running the latest version
>
```

**Relevant plugins (Delete if this isn't needed):**
only happens when CMI is installed.

**Comment by mrfloris** [comment]
I believe from kits this is now fixed, did you have time to do that book-fix-on-use thing yet?

**Comment by LogGits** [comment]
This is a minecraft bug fixed in 1.16. It may be worth contacting the paper or spigot developers for a temp fix. 

https://bugs.mojang.com/browse/MC-179868
